account-data-matching and bump-seed-canonicalization lesson updates (#353)
solana-foundation/developer-content: content/courses/program-security/account-data-matching.md

title: Account Data Matching
objectives:
  - Explain the security risks associated with missing data validation checks
  - Implement data validation checks using native Rust
  - Implement data validation checks using Anchor constraints
description:
  Learn how to properly validate account data in Solana programs to prevent
  security vulnerabilities.
---

# Account Data Matching

## Summary

- **Data validation checks** are crucial for verifying that account data matches
  expected values. Without proper checks, malicious users could exploit your
  program using unexpected accounts.
- Implement data validation in native Rust by comparing account data to expected

>   values:

Reviewer: Much better than previous.

Author: I agree, emphasis helped summarize!

  ```rust
  if ctx.accounts.user.key() != ctx.accounts.user_data.user {
      return Err(ProgramError::InvalidAccountData.into());
  }
  ```

> - Use Anchor constraints to simplify the process:

Reviewer: ```suggestion
- Use [Anchor constraints](https://www.anchor-lang.com/docs/account-constraints) to simplify the process:
```

  - Use `constraint` to evaluate custom expressions
  - Use `has_one` to check that a field on one account matches the key of
    another account

## Lesson

> Account data matching is a critical security practice in Solana program
> development. It involves implementing data validation checks to ensure that the
> data stored on an account matches expected values. These checks are essential
> for preventing unauthorized access and maintaining the integrity of your
> program's state.

Reviewer: ```suggestion
Account data matching involves implementing data validation checks to ensure that the
data stored on an account matches expected values. This is essential for preventing 
unauthorized access and maintaining the integrity of your program's state.
```

Reviewer: This was good, just a little repetitive.

Author: Got it!

### Why Account Data Matching Matters

Without proper data validation, your program becomes vulnerable to various
attacks:

> 1. Unauthorized access: Malicious users could pass in unexpected accounts,
>    gaining access to functionality they shouldn't have
> 2. State manipulation: Attackers might alter the program's state in unintended
>    ways, compromising its integrity
> 3. Fund theft: In programs dealing with tokens or SOL, inadequate checks could
>    lead to unauthorized withdrawals

Reviewer: ```suggestion
1. Unauthorized access: malicious users could pass in unexpected accounts,
   gaining access to functionality they shouldn't have
2. State manipulation: attackers might alter the program's state in unintended
   ways, compromising its integrity
3. Fund theft: in programs dealing with tokens or SOL, inadequate checks could
   lead to unauthorized withdrawals
```

}
```

This instruction is insecure because it doesn't verify that the `current_admin`
signer matches the `admin` stored in the `admin_config` account. Any signer
could potentially update the admin!

### Implementing Data Validation Checks

Let's secure this instruction using two methods: native Rust and Anchor
constraints.

#### Native Rust Approach

In native Rust, we can add a simple comparison:

```rust
pub fn update_admin(ctx: Context<UpdateAdmin>) -> Result<()> {
    if ctx.accounts.current_admin.key() != ctx.accounts.admin_config.admin {
        return Err(ProgramError::InvalidAccountData.into());
    }
    ctx.accounts.admin_config.admin = ctx.accounts.new_admin.key();
    Ok(())
}
```

This check ensures that only the current admin can update the admin_config
account.

#### Anchor Constraints

> Anchor provides a more declarative way to implement these checks using
> constraints:

Reviewer: ```suggestion
Anchor provides a more declarative way to implement these checks using
[account constraints](https://www.anchor-lang.com/docs/account-constraints):
```

Author: Smart!

    )]
    pub admin_config: Account<'info, AdminConfig>,
    pub current_admin: Signer<'info>,
    /// CHECK: This account is not read or written in this instruction
    pub new_admin: UncheckedAccount<'info>,
}
```

The `has_one = current_admin` constraint checks that the `current_admin`
account's key matches the `admin` field in the `AdminConfig` account. If it
doesn't match, Anchor will return the custom `InvalidAdmin` error.

Alternatively, you can use the `constraint` attribute for more complex checks:

```rust
#[derive(Accounts)]
pub struct UpdateAdmin<'info> {
    #[account(
        mut,
        constraint = admin_config.admin == current_admin.key() @ MyError::InvalidAdmin
    )]
    pub admin_config: Account<'info, AdminConfig>,
    pub current_admin: Signer<'info>,
    /// CHECK: This account is not read or written in this instruction
    pub new_admin: UncheckedAccount<'info>,
}
```

This approach allows for more flexibility in your validation logic.

<Callout type="info">
  Remember to define your custom errors in your program's error enum:

```rust
#[error_code]
pub enum MyError {
    #[msg("Only the current admin can perform this action")]
    InvalidAdmin,
}
```

> </Callout>

Reviewer: Thanks @Epistetechnician ! I appreciate you actually read the CONTRIBUTING guide.

Author: This is the way ser!

By implementing these checks, you ensure that only the rightful admin can update

> the admin_config account, significantly improving your program's security.

Reviewer: ```suggestion
the `admin_config` account, significantly improving your program's security.
```

## Lab: Securing a Vault Program

In this lab, we'll create a simple "vault" program to demonstrate the importance
of account data matching. We'll implement both an insecure and a secure version
of a withdraw instruction to highlight the difference.

### 1. Setup

Clone the starter code from the `starter` branch of

> [this repository](https://github.com/Unboxed-Software/solana-account-data-matching).

Reviewer: Did this need repo need any updates @Epistetechnician ? It's 2 years old - per CONTRIBUTING we'd expect a PR on the link repository too.

Author: Yes for sure! Shall I get started on that?

Reviewer: Yes please do!

Reviewer: I've cloned the repo to https://github.com/solana-developers/solana-account-data-matching and invited you as a contributor. Please update the repo then once complete we can get this PR in!

The starter code includes a program with two instructions and a boilerplate test
file:

> 1. `initialize_vault`: Creates a new `Vault` account and associated
>    `TokenAccount`.

Reviewer: ```suggestion
1. `initialize_vault`: Creates a new `Vault` account and related `TokenAccount`.
```

Want to avoid "associated `TokenAccount`" because we already have "associated token accounts" and my understanding is that not what is meant here. Please let me know if that's not correct.

Author: You are right my apologies!

    pub token_program: Program<'info, Token>,
    pub authority: Signer<'info>,
}
```

The critical flaw here is that while the instruction requires an `authority`
signer, it doesn't verify whether the `authority` account passed to the
instruction matches the `authority` stored in the `Vault` account. This means
that any signer could potentially withdraw funds from the vault, even if they're
not the authorized user.

### 3. Test the Insecure Withdraw Instruction

Let's write a test to demonstrate this vulnerability. Add the following test to
your `tests/account-data-matching.ts` file:

```typescript

> it("Insecure withdraw allows unauthorized access", async () => {

Reviewer: This is very minor, but you're combining TSS/BDD style in your tests. It's super common everyone does it. `it` is supposed to read a little like a sentence, where `it` is the thing in the `describe` block, that's why it displays that way on the command line.

You should either make this `test("Insecure withdraw allows unauthorized access" ... (probably simplest) ` or `describe('insecure withdraw)` ....`it("allows unauthorized access" ... ` 

Do that for all the tests.

}
```

> The key difference here is the use of `has_one` constraints in the
> `SecureWithdraw` struct. These ensure that the `authority`, `token_account`, and

Reviewer: ```suggestion
The key difference here is the use of a [`has_one` constraint](https://www.anchor-lang.com/docs/account-constraints) in the `SecureWithdraw` struct. These ensure that the `authority`, `token_account`, and
```

`withdraw_destination` passed to the instruction match those stored in the
`Vault` account.

### 5. Test the Secure Withdraw Instruction

Now, let's test our secure instruction:

```typescript
it("Secure withdraw prevents unauthorized access", async () => {
  // Setup: Initialize vault and mint some tokens to it
  // ... (initialization code here)

  // Attempt unauthorized withdrawal

>   // Note: walletFake represents an unauthorized wallet trying to withdraw funds

Reviewer: This is one of those comments that indicate you have found a better name for something.

Rename `walletFake` to `unauthorizedWallet`

Rename `withdrawDestinationFake` to `unauthorizedWalletWithdrawDestination` or similar.

Author: Understood!

      tokenAccount: tokenPDA,
      withdrawDestination: withdrawDestination,
      authority: wallet.publicKey,
    })
    .rpc();

  // Check that the withdrawal succeeded
  const balance = await connection.getTokenAccountBalance(tokenPDA);
  expect(balance.value.uiAmount).to.eq(0);
});
```

Run `anchor test` again. You should see that the unauthorized withdrawal now
fails, while the authorized one succeeds.

### Conclusion

By implementing proper account data matching, we've significantly improved the
security of our vault program. The secure version ensures that only the
authorized user can withdraw funds, preventing potential exploits.

Remember, as your programs grow in complexity, it becomes increasingly important
to implement thorough data validation checks. Always consider what assumptions
your program is making about its inputs. Validate these assumptions explicitly.

Now that you've seen how to implement account data matching, take some time to
review one of your existing programs. Look for places where you might be making
assumptions about account data without explicitly checking it. Implement
appropriate checks using the techniques you've learned in this lesson.

> Remember, if you find a security vulnerability in someone else's program,
> responsibly disclose it to the program's maintainers. If you find one in your
> own program, patch it as soon as possible!

Reviewer: ```suggestion
Remember, if you find a security vulnerability in someone else's program,
[responsibly disclose](https://en.wikipedia.org/wiki/Coordinated_vulnerability_disclosure)
it to the program's maintainers. If you find one in your own program, patch it as soon as possible!
```
